File: scripts/modes/gameMode.js
// Abstract base class for all game modes
export class GameMode {
    constructor(game) {
        this.game = game;
        this.name = 'Base Mode';
        this.description = 'Base game mode';
        this.objectives = [];
        this.modeSpecificStats = {};
        this.isComplete = false;
        this.isPaused = false;
        this.startTime = null;
        this.endTime = null;
    }

    // Abstract methods that must be implemented by subclasses
    initialize() {
        throw new Error('initialize() must be implemented by subclass');
    }

    update(deltaTime) {
        throw new Error('update() must be implemented by subclass');
    }

    handleLineClears(linesCleared, specialClear) {
        throw new Error('handleLineClears() must be implemented by subclass');
    }

    getObjective() {
        throw new Error('getObjective() must be implemented by subclass');
    }

    getModeUI() {
        throw new Error('getModeUI() must be implemented by subclass');
    }

    // Common methods that can be overridden if needed
    start() {
        this.startTime = Date.now();
        this.isComplete = false;
        this.isPaused = false;
        this.initialize();
    }

    pause() {
        this.isPaused = true;
    }

    resume() {
        this.isPaused = false;
    }

    end(won = false) {
        this.endTime = Date.now();
        this.isComplete = true;
        const duration = this.endTime - this.startTime;
        return {
            mode: this.name,
            won: won,
            duration: duration,
            stats: this.getModeStats()
        };
    }

    getModeStats() {
        return {
            ...this.modeSpecificStats,
            duration: this.endTime ? this.endTime - this.startTime : Date.now() - this.startTime
        };
    }

    // Save and load state for offline support
    saveState() {
        return {
            name: this.name,
            startTime: this.startTime,
            modeSpecificStats: this.modeSpecificStats,
            isComplete: this.isComplete,
            isPaused: this.isPaused
        };
    }

    loadState(state) {
        this.startTime = state.startTime;
        this.modeSpecificStats = state.modeSpecificStats;
        this.isComplete = state.isComplete;
        this.isPaused = state.isPaused;
    }

    // Check if mode has special rules
    hasSpecialRules() {
        return false;
    }

    getSpecialRules() {
        return {};
    }

    // Handle mode-specific input
    handleInput(action) {
        // Override in subclasses if needed
        return false;
    }

    // Get current progress percentage
    getProgress() {
        return 0;
    }

    // Check if the mode should end
    shouldEnd() {
        return false;
    }

    // Get mode-specific scoring multiplier
    getScoreMultiplier() {
        return 1;
    }

    // Get mode-specific speed level
    getSpeedLevel() {
        return this.game.level;
    }

    // Check if hold piece is allowed
    isHoldAllowed() {
        return true;
    }

    // Check if ghost piece should be shown
    showGhostPiece() {
        return true;
    }

    // Get mode-specific background music tempo
    getMusicTempo() {
        return 1;
    }

    // Handle game over condition
    handleGameOver() {
        return this.end(false);
    }

    // Handle victory condition
    handleVictory() {
        return this.end(true);
    }

    // Render mode-specific UI elements
    renderModeSpecificUI(ctx, canvas) {
        // Override in subclasses to render custom UI
    }

    // Get leaderboard category for this mode
    getLeaderboardCategory() {
        return this.name.toLowerCase().replace(/\s+/g, '_');
    }

    // Check if mode supports saving mid-game
    supportsSaving() {
        return false;
    }

    // Get mode icon for UI
    getIcon() {
        return '🎮';
    }

    // Get mode color theme
    getThemeColor() {
        return 'var(--neon-blue)';
    }
}

File: android-app/app/src/main/java/com/tetris/modern/rl/game/modes/GameMode.kt
package com.tetris.modern.rl.game.modes

import com.tetris.modern.rl.game.TetrisEngine
import com.tetris.modern.rl.game.models.GameState

/**
 * Abstract base class for all game modes
 */
abstract class GameMode(val name: String) {
    
    protected var engine: TetrisEngine? = null
    
    /**
     * Initialize the game mode with the engine
     */
    open fun initialize(engine: TetrisEngine) {
        this.engine = engine
    }
    
    /**
     * Update the game mode logic
     */
    abstract fun update(deltaTime: Long, gameState: GameState)
    
    /**
     * Handle line clears specific to this mode
     */
    abstract fun handleLineClears(lines: Int, isTSpin: Boolean)
    
    /**
     * Get the objective text for this mode
     */
    abstract fun getObjective(): String
    
    /**
     * Get mode-specific UI elements
     */
    open fun getModeUI(): Map<String, String> {
        return emptyMap()
    }
    
    /**
     * Check if the mode has a win condition and if it's met
     */
    open fun checkWinCondition(gameState: GameState): Boolean {
        return false
    }
    
    /**
     * Handle game over for this mode
     */
    open fun onGameOver(gameState: GameState) {
        // Override in subclasses for mode-specific game over handling
    }
    
    /**
     * Get the description of this mode
     */
    abstract fun getDescription(): String
    
    /**
     * Check if this mode is available on mobile
     */
    open fun isAvailableOnMobile(): Boolean {
        return true
    }
}